src/megaapi_impl.cpp

```diff
@@ -12630,30 +12717,41 @@ void MegaApiImpl::folderlinkinfo_result(error e, handle owner, handle /*ph*/, st
 }
 
 #ifdef ENABLE_SYNC
-void MegaApiImpl::syncupdate_state(Sync *sync, syncstate_t newstate)
+void MegaApiImpl::syncupdate_state(int tag, syncstate_t newstate, syncerror_t syncerror, bool fireDisableEvent)
 {
-    if(syncMap.find(sync->tag) == syncMap.end()) return;
-    MegaSyncPrivate* megaSync = syncMap.at(sync->tag);
+    auto syncpair = syncMap.find(tag);
+    if( syncpair == syncMap.end())
+    {
+        LOG_err << " updating state for missing sync: tag = " << tag << " newstate = " << newstate << " err = " << syncerror;
+        return;
+    }
+    MegaSyncPrivate* megaSync = syncpair->second;
+
+    auto previousState = megaSync->getState();
+    bool wasActive = megaSync->isActive(); // Beware: new sync is active (constructor set state to INITIAL_SCAN)
+    bool wasEnabled = megaSync->isEnabled();
+    bool wasTemporaryDisabled = megaSync->isTemporaryDisabled();
     megaSync->setState(newstate);
-    LOG_debug << "Sync state change: " << newstate << " Path: " << sync->localroot->name;
-    client->abortbackoff(false);
+    megaSync->setError(syncerror);
+    LOG_debug << "Sync state change: " << newstate << " syncError: " << syncerror << " Path: " << megaSync->getLocalFolder();
+    LOG_verbose << "Sync previous state change: " << previousState << " wasActive: " << wasActive << " wasEnabled: " << wasEnabled << " Path: " << megaSync->getLocalFolder();
 
-    if (newstate == SYNC_FAILED)
-    {
-        MegaRequestPrivate *request = new MegaRequestPrivate(MegaRequest::TYPE_ADD_SYNC);
+    fireOnSyncStateChanged(megaSync);
 
-        if(sync->localroot->node)
-        {
-            request->setNodeHandle(sync->localroot->node->nodehandle);
-        }
+    if (fireDisableEvent && ( (wasActive && !megaSync->isActive()) || (wasTemporaryDisabled && !megaSync->isEnabled()) ) )
```

Comment:
no. why?

Reply:
Because I was wondering why you declare variables to store the `bool wasEnabled = megaSync->isEnabled()` and alike, but after `fireOnSyncStateChanged()`, you check `!megaSync->isEnabled()` instead of using the variable `wasEnabled` directly. Same applies to `wasActive`

Reply:
the key are:
```
    megaSync->setState(newstate);
    megaSync->setError(syncerror);
```
which change the state of the MegaSync object. I'll add some comments for better reading ;-)